File: app/src/main/java/com/xeniac/warrantyroster/mainactivity/warrantiesfragment/WarrantiesFragment.java
package com.xeniac.warrantyroster.mainactivity.warrantiesfragment;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import com.xeniac.warrantyroster.R;
import com.xeniac.warrantyroster.databinding.FragmentWarrantiesBinding;

public class WarrantiesFragment extends Fragment implements WarrantyListClickInterface {

    private FragmentWarrantiesBinding warrantiesBinding;
    private View view;
    private Activity activity;
    private Context context;
    private NavController navController;

    public WarrantiesFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        warrantiesBinding = FragmentWarrantiesBinding.inflate(inflater, container, false);
        view = warrantiesBinding.getRoot();
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        warrantiesBinding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        activity = getActivity();
        context = getContext();
        navController = Navigation.findNavController(view);

        boolean isListEmpty = false;
        //Todo can use databinding
        if (isListEmpty) {
            warrantiesBinding.searchWarranties.setVisibility(View.GONE);
            warrantiesBinding.rvWarranties.setVisibility(View.GONE);
            warrantiesBinding.groupWarrantiesEmptyList.setVisibility(View.VISIBLE);
        } else {
            warrantiesBinding.searchWarranties.setVisibility(View.VISIBLE);
            warrantiesBinding.groupWarrantiesEmptyList.setVisibility(View.GONE);
            warrantiesBinding.rvWarranties.setVisibility(View.VISIBLE);
            showWarrantyList();
            search();
        }
    }

    private void showWarrantyList() {
        WarrantyAdapter warrantyAdapter = new WarrantyAdapter(context, WarrantyDataProvider.warrantyList, this);
        warrantiesBinding.rvWarranties.setAdapter(warrantyAdapter);
    }

    @Override
    public void onItemClick(int position) {
        Toast.makeText(context, WarrantyDataProvider.warrantyList.get(position).getTitle() + " clicked.",
                Toast.LENGTH_SHORT).show();
    }

    private void search() {
        warrantiesBinding.searchWarranties.setOnQueryTextFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean hasFocus) {
                if (hasFocus) {
                    warrantiesBinding.toolbarWarranties.setTitle(null);
                } else {
                    warrantiesBinding.toolbarWarranties.setTitle(context.getResources().getString(R.string.warranties_text_title));
                }
            }
        });

        warrantiesBinding.searchWarranties.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                if (!TextUtils.isEmpty(query)) {
                    Toast.makeText(context, "onQueryTextSubmit", Toast.LENGTH_SHORT).show();
                    warrantiesBinding.searchWarranties.onActionViewCollapsed();
                }
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (!TextUtils.isEmpty(newText)) {
                    Toast.makeText(context, "Input: " + newText, Toast.LENGTH_SHORT).show();
                }
                return false;
            }
        });
    }
}

File: app/src/main/java/com/xeniac/warrantyroster_manager/ui/main/fragments/WarrantiesFragment.kt
package com.xeniac.warrantyroster_manager.ui.main.fragments

import android.os.Bundle
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import com.google.android.material.snackbar.Snackbar
import com.xeniac.warrantyroster_manager.R
import com.xeniac.warrantyroster_manager.data.remote.models.Warranty
import com.xeniac.warrantyroster_manager.databinding.FragmentWarrantiesBinding
import com.xeniac.warrantyroster_manager.ui.main.adapters.WarrantyAdapter
import com.xeniac.warrantyroster_manager.ui.main.adapters.WarrantyListClickInterface
import com.xeniac.warrantyroster_manager.ui.viewmodels.MainViewModel
import com.xeniac.warrantyroster_manager.utils.Constants.ERROR_EMPTY_CATEGORY_LIST
import com.xeniac.warrantyroster_manager.utils.Constants.ERROR_EMPTY_WARRANTY_LIST
import com.xeniac.warrantyroster_manager.utils.Constants.ERROR_FIREBASE_403
import com.xeniac.warrantyroster_manager.utils.Constants.ERROR_FIREBASE_DEVICE_BLOCKED
import com.xeniac.warrantyroster_manager.utils.SnackBarHelper.showFirebaseDeviceBlockedError
import com.xeniac.warrantyroster_manager.utils.Status
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class WarrantiesFragment : Fragment(R.layout.fragment_warranties), WarrantyListClickInterface {

    private var _binding: FragmentWarrantiesBinding? = null
    val binding get() = _binding!!

    lateinit var viewModel: MainViewModel

    @Inject
    lateinit var warrantyAdapter: WarrantyAdapter

    private var snackbar: Snackbar? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        _binding = FragmentWarrantiesBinding.bind(view)
        viewModel = ViewModelProvider(requireActivity())[MainViewModel::class.java]

        getCategoriesFromFirestore()
        setupRecyclerView()
        subscribeToObservers()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        snackbar?.dismiss()
        _binding = null
    }

    private fun setupRecyclerView() {
        warrantyAdapter.apply {
            setOnWarrantyItemClickListenter(this@WarrantiesFragment)
            activity = requireActivity()
            context = requireContext()
            mainViewModel = viewModel
        }
        binding.rv.adapter = warrantyAdapter
    }

    private fun subscribeToObservers() {
        categoriesListObserver()
        warrantiesListObserver()
    }

    private fun getCategoriesFromFirestore() = viewModel.getCategoriesFromFirestore()

    private fun getWarrantiesListFromFirestore() = viewModel.getWarrantiesListFromFirestore()

    private fun categoriesListObserver() {
        viewModel.categoriesLiveData.observe(viewLifecycleOwner) { responseEvent ->
            responseEvent.peekContent().let { response ->
                when (response.status) {
                    Status.LOADING -> showLoadingAnimation()
                    Status.SUCCESS -> getWarrantiesListFromFirestore()
                    Status.ERROR -> {
                        response.message?.let {
                            when {
                                it.contains(ERROR_EMPTY_CATEGORY_LIST) -> {
                                    getCategoriesFromFirestore()
                                }
                                it.contains(ERROR_FIREBASE_403) -> {
                                    binding.tvNetworkError.text =
                                        requireContext().getString(R.string.error_firebase_403)
                                    showNetworkError()
                                }
                                it.contains(ERROR_FIREBASE_DEVICE_BLOCKED) -> {
                                    snackbar = showFirebaseDeviceBlockedError(
                                        requireContext(), binding.root
                                    )
                                }
                                else -> {
                                    binding.tvNetworkError.text =
                                        requireContext().getString(R.string.error_network_connection)
                                    showNetworkError()
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private fun warrantiesListObserver() {
        viewModel.warrantiesLiveData.observe(viewLifecycleOwner) { responseEvent ->
            responseEvent.getContentIfNotHandled()?.let { response ->
                when (response.status) {
                    Status.LOADING -> showLoadingAnimation()
                    Status.SUCCESS -> {
                        hideLoadingAnimation()
                        response.data?.let { warrantiesList ->
                            showWarrantiesList(warrantiesList)
                        }
                    }
                    Status.ERROR -> {
                        hideLoadingAnimation()
                        response.message?.let {
                            when {
                                it.contains(ERROR_EMPTY_WARRANTY_LIST) -> {
                                    showWarrantiesEmptyList()
                                }
                                it.contains(ERROR_FIREBASE_403) -> {
                                    binding.tvNetworkError.text =
                                        requireContext().getString(R.string.error_firebase_403)
                                    showNetworkError()
                                }
                                it.contains(ERROR_FIREBASE_DEVICE_BLOCKED) -> {
                                    snackbar = showFirebaseDeviceBlockedError(
                                        requireContext(), binding.root
                                    )
                                }
                                else -> {
                                    binding.tvNetworkError.text =
                                        requireContext().getString(R.string.error_network_connection)
                                    showNetworkError()
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private fun showNetworkError() {
        binding.groupEmptyList.visibility = GONE
        binding.rv.visibility = GONE
        binding.groupNetwork.visibility = VISIBLE
        retryNetworkBtn()
    }

    private fun retryNetworkBtn() = binding.btnNetworkRetry.setOnClickListener {
        getWarrantiesListFromFirestore()
    }

    private fun showLoadingAnimation() {
        binding.svWarranties.visibility = GONE
        binding.groupNetwork.visibility = GONE
        binding.groupEmptyList.visibility = GONE
        binding.rv.visibility = GONE
        binding.cpi.visibility = VISIBLE
        binding.cpi.show()
    }

    private fun hideLoadingAnimation() {
        binding.cpi.hide()
        binding.cpi.setVisibilityAfterHide(GONE)
    }

    private fun showWarrantiesEmptyList() {
        binding.svWarranties.visibility = GONE
        binding.groupNetwork.visibility = GONE
        binding.rv.visibility = GONE
        binding.groupEmptyList.visibility = VISIBLE
    }

    private fun showWarrantiesList(warrantiesList: MutableList<Warranty>) {
        binding.groupNetwork.visibility = GONE
        binding.groupEmptyList.visibility = GONE
        binding.rv.visibility = VISIBLE
        warrantyAdapter.warrantiesList = warrantiesList

        //TODO remove comment after adding search function
//        searchWarrantiesList();
    }

    override fun onItemClick(warranty: Warranty) = findNavController().navigate(
        WarrantiesFragmentDirections.actionWarrantiesFragmentToWarrantyDetailsFragment(warranty)
    )

    /*
    private fun searchWarrantiesList() {
        binding.svWarranties.setVisibility(VISIBLE);

        binding.svWarranties.setOnQueryTextFocusChangeListener((view, hasFocus) -> {
            if (hasFocus) {
                warrantiesBinding.toolbarWarranties.setTitle(null);
            } else {
                warrantiesBinding.toolbarWarranties.setTitle(
                    context.getResources().getString(R.string.warranties_text_title)
                );
            }
        });

        warrantiesBinding.svWarranties.setOnQueryTextListener(new SearchView . OnQueryTextListener () {
            @Override
            public boolean onQueryTextSubmit(String query) {
                if (!TextUtils.isEmpty(query)) {
                    Toast.makeText(context, "onQueryTextSubmit", Toast.LENGTH_SHORT).show();
                    warrantiesBinding.svWarranties.onActionViewCollapsed();
                }
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (!TextUtils.isEmpty(newText)) {
                    Toast.makeText(context, "Input: " + newText, Toast.LENGTH_SHORT).show();
                }
                return false;
            }
        });
    }
     */
}